pgx/src/subxact.rs

```diff
@@ -0,0 +1,196 @@
+use crate::{pg_sys, PgMemoryContexts, SpiClient};
+use std::fmt::{Debug, Formatter};
+use std::ops::Deref;
+
+/// Sub-transaction
+///
+/// Can be created by calling `SpiClient::sub_transaction`, `SubTransaction<Parent>::sub_transaction`
+/// or any other implementation of `SubTransactionExt` and obtaining it as an argument to the provided closure.
+///
+/// Unless rolled back or committed explicitly, it'll commit if `COMMIT` generic parameter is `true`
+/// (default) or roll back if it is `false`.
+pub struct SubTransaction<Parent: SubTransactionExt, const COMMIT: bool = true> {
+    memory_context: pg_sys::MemoryContext,
+    resource_owner: pg_sys::ResourceOwner,
+    // Should the transaction be released, or was it already committed or rolled back?
+    //
+    // The reason we are not calling this `released` as we're also using this flag when
+    // we convert between commit_on_drop and rollback_on_drop to ensure it doesn't get released
+    // on the drop of the original value.
+    should_release: bool,
+    parent: Option<Parent>,
+}
+
+impl<Parent: SubTransactionExt, const COMMIT: bool> Debug for SubTransaction<Parent, COMMIT> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(std::any::type_name::<Self>())
+    }
+}
+
+impl<Parent: SubTransactionExt, const COMMIT: bool> SubTransaction<Parent, COMMIT> {
+    /// Create a new sub-transaction.
+    fn new(parent: Parent) -> Self {
+        // Remember the memory context before starting the sub-transaction
+        let ctx = PgMemoryContexts::CurrentMemoryContext.value();
+        // Remember resource owner before starting the sub-transaction
+        let resource_owner = unsafe { pg_sys::CurrentResourceOwner };
+        unsafe {
+            pg_sys::BeginInternalSubTransaction(std::ptr::null());
+        }
+        // Switch to the outer memory context so that all allocations remain
+        // there instead of the sub-transaction's context
+        PgMemoryContexts::For(ctx).set_as_current();
+        Self { memory_context: ctx, should_release: true, resource_owner, parent: Some(parent) }
+    }
+
+    /// Commit the transaction, returning its parent
+    pub fn commit(mut self) -> Parent {
+        self.internal_commit();
+        self.should_release = false;
+        self.parent.take().unwrap()
+    }
+
+    /// Rollback the transaction, returning its parent
+    pub fn rollback(mut self) -> Parent {
+        self.internal_rollback();
+        self.should_release = false;
+        self.parent.take().unwrap()
+    }
```

Comment:
It seems this code is expecting `self.parent` to actually be unconditionally `Some`, so why is it that we use an `Option`? Just for `mem::take` antics later?

Reply:
IIRC, this was because I couldn't move it out of the `SubTransaction` when I am changing the COMMIT type parameter because it implements `Drop`.

Reply:
Ultimately, I wouldn't mind getting rid of `Option` (I tried!) but I haven't thought of anything better yet.

Reply:
You can split `SubTransaction` into two fields. You can't destructure types that have `impl Drop`, but you can destructure types made of types that `impl Drop`, so you move the feature of the commit-on-drop (and thus thread the const generic) into that inner field. Since users can't access the type's inner workings, it still has `impl Drop` as far as they are concerned.

Then you use destructuring syntax to get at the fields, and instead of using a `should_release` bool, you simply `mem::forget` it. This also lets you elevate the requirements for `internal_commit` and `internal_rollback`.